postgis/lwgeom_remove_small_parts.c

```diff
@@ -0,0 +1,244 @@
+/**********************************************************************
+ *
+ * PostGIS - Spatial Types for PostgreSQL
+ * http://postgis.net
+ *
+ * PostGIS is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright (C) 2024 Sam Peters <gluser1357@gmx.de>
+ *
+ **********************************************************************/
+
+#include "postgres.h"
+#include "funcapi.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/numeric.h"
+#include "access/htup_details.h"
+
+#include "liblwgeom.h"
+#include "liblwgeom_internal.h"
+
+// ===============================================================================
+// helper function for polygon and polyline POINTARRAY's
+// which removes small parts given by dx and dy.
+// ===============================================================================
+static void ptarray_remove_dim_helper(POINTARRAY *points, double mindx, double mindy) {
+
+    int r;
+    double xmin = 0, ymin = 0, xmax = 0, ymax = 0;
```

Comment:
We could do an early check for `points->npoints == 0`, then initialize these to the coordinates of the first point. Or maybe even use an existing envelope function.

Reply:
Do you mean like the following? IMO the disadvantage is +5 lines of (repeated) code, not sure if this would make the code more readable or faster.


    int r;
    double xmin, ymin, xmax, ymax;
    double x, y;
    POINT4D point;

    if (points->npoints == 0) return;

    getPoint4d_p(points, 0, &point); // point read/write see ptarray_flip_coordinates
    xmin = xmax = point.x;
    ymin = ymax = point.y;
     
    for (r=1; r < points->npoints; r++) {

		getPoint4d_p(points, r, &point); // point read/write see ptarray_flip_coordinates

		x = point.x;
		y = point.y;

		if (mindx > 0) {
		    if (xmin > x) xmin = x;
		    if (xmax < x) xmax = x;
		}
		if (mindy > 0) {
		    if (ymin > y) ymin = y;
		    if (ymax < y) ymax = y;
		}
	}

Reply:
Ok then I leave it for now. Thank you anyway for the hint :)